data/transactions/transaction.go

```diff
@@ -288,323 +286,90 @@ func (tx Header) TxFee() basics.MicroAlgos {
 func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusParams) error {
 	switch tx.Type {
 	case protocol.PaymentTx:
-		// in case that the fee sink is spending, check that this spend is to a valid address
-		err := tx.checkSpender(tx.Header, spec, proto)
+		err := tx.PaymentTxnFields.wellFormed(tx.Header, spec, proto)
 		if err != nil {
 			return err
 		}
 
 	case protocol.KeyRegistrationTx:
-		if proto.EnableKeyregCoherencyCheck {
-			// ensure that the VoteLast is greater or equal to the VoteFirst
-			if tx.KeyregTxnFields.VoteFirst > tx.KeyregTxnFields.VoteLast {
-				return errKeyregTxnFirstVotingRoundGreaterThanLastVotingRound
-			}
-
-			// The trio of [VotePK, SelectionPK, VoteKeyDilution] needs to be all zeros or all non-zero for the transaction to be valid.
-			if !((tx.KeyregTxnFields.VotePK.IsEmpty() && tx.KeyregTxnFields.SelectionPK.IsEmpty() && tx.KeyregTxnFields.VoteKeyDilution == 0) ||
-				(!tx.KeyregTxnFields.VotePK.IsEmpty() && !tx.KeyregTxnFields.SelectionPK.IsEmpty() && tx.KeyregTxnFields.VoteKeyDilution != 0)) {
-				return errKeyregTxnNonCoherentVotingKeys
-			}
-
-			// if it's a going offline transaction
-			if tx.KeyregTxnFields.VoteKeyDilution == 0 {
-				// check that we don't have any VoteFirst/VoteLast fields.
-				if tx.KeyregTxnFields.VoteFirst != 0 || tx.KeyregTxnFields.VoteLast != 0 {
-					return errKeyregTxnOfflineTransactionHasVotingRounds
-				}
-			} else {
-				// going online
-				if tx.KeyregTxnFields.VoteLast == 0 {
-					return errKeyregTxnGoingOnlineWithZeroVoteLast
-				}
-				if tx.KeyregTxnFields.VoteFirst > tx.LastValid+1 {
-					return errKeyregTxnGoingOnlineWithFirstVoteAfterLastValid
-				}
-			}
-		}
-
-		// check that, if this tx is marking an account nonparticipating,
-		// it supplies no key (as though it were trying to go offline)
-		if tx.KeyregTxnFields.Nonparticipation {
-			if !proto.SupportBecomeNonParticipatingTransactions {
-				// if the transaction has the Nonparticipation flag high, but the protocol does not support
-				// that type of transaction, it is invalid.
-				return errKeyregTxnUnsupportedSwitchToNonParticipating
-			}
-			suppliesNullKeys := tx.KeyregTxnFields.VotePK.IsEmpty() || tx.KeyregTxnFields.SelectionPK.IsEmpty()
-			if !suppliesNullKeys {
-				return errKeyregTxnGoingOnlineWithNonParticipating
-			}
-		}
-
-		if err := tx.stateProofPKWellFormed(proto); err != nil {
+		err := tx.KeyregTxnFields.wellFormed(tx.Header, spec, proto)
+		if err != nil {
 			return err
 		}
 
-	case protocol.AssetConfigTx:
+	case protocol.AssetConfigTx, protocol.AssetTransferTx, protocol.AssetFreezeTx:
 		if !proto.Asset {
 			return fmt.Errorf("asset transaction not supported")
 		}
 
-	case protocol.AssetTransferTx:
-		if !proto.Asset {
-			return fmt.Errorf("asset transaction not supported")
-		}
-
-	case protocol.AssetFreezeTx:
-		if !proto.Asset {
-			return fmt.Errorf("asset transaction not supported")
-		}
 	case protocol.ApplicationCallTx:
```

Comment:
I think you could have chosen to just seperate out the App call validation and accomplished most of the goal!

Reply:
Yeah, maybe, but it would have been grossly asymmetric!